File: android/app/src/main/java/com/applens/processor/PatternDetector.kt
package com.applens.processor

/**
 * Detect component patterns in a parsed screen tree.
 * Tags screens based on the UI components present.
 */
object PatternDetector {

    data class DetectionResult(
        val patterns: List<String>,
        val componentCounts: Map<String, Int>,
        val summary: String,
        val totalNodes: Int
    )

    fun detect(tree: ScreenTree): DetectionResult {
        if (tree.root == null) return DetectionResult(emptyList(), emptyMap(), "Empty screen", 0)

        val patterns = mutableSetOf<String>()
        val componentCounts = mutableMapOf<String, Int>()

        fun walk(node: ComponentNode) {
            val cls = node.className.lowercase()
            val shortName = XmlProcessor.getClassShortName(node.className)
            componentCounts[shortName] = (componentCounts[shortName] ?: 0) + 1

            when {
                cls.contains("recyclerview") || cls.contains("listview") -> patterns.add("List")
                cls.contains("edittext") || cls.contains("textinputedittext") || cls.contains("textinput") -> patterns.add("Input Form")
                cls.contains("bottomnavigationview") || cls.contains("bottomnavigation") -> patterns.add("Bottom Nav")
                cls.contains("navigationview") || cls.contains("drawerlayout") -> patterns.add("Navigation Drawer")
                cls.contains("webview") -> patterns.add("WebView")
                cls.contains("viewpager") || cls.contains("viewpager2") -> patterns.add("ViewPager")
                cls.contains("tablayout") || cls.contains("tabhost") -> patterns.add("Tab Layout")
                cls.contains("toolbar") || cls.contains("appbar") || cls.contains("actionbar") -> patterns.add("Toolbar/AppBar")
                cls.contains("floatingactionbutton") || cls.contains("fab") -> patterns.add("FAB")
                cls.contains("imageview") || cls.contains("image") -> patterns.add("Images")
                cls.contains("checkbox") -> patterns.add("Checkbox")
                cls.contains("radiobutton") -> patterns.add("Radio Button")
                cls.contains("switch") || cls.contains("togglebutton") -> patterns.add("Switch/Toggle")
                cls.contains("spinner") || cls.contains("dropdown") -> patterns.add("Dropdown/Spinner")
                cls.contains("searchview") || cls.contains("searchbar") -> patterns.add("Search Bar")
                cls.contains("datepicker") || cls.contains("timepicker") || cls.contains("calendar") -> patterns.add("Date/Time Picker")
                cls.contains("video") || cls.contains("exoplayer") || cls.contains("mediaplayer") -> patterns.add("Media Player")
                cls.contains("dialog") || cls.contains("alertdialog") || cls.contains("modal") -> patterns.add("Dialog")
                cls.contains("progressbar") || cls.contains("loading") -> patterns.add("Progress/Loading")
            }

            for (child in node.children) walk(child)
        }

        walk(tree.root)

        val topComponents = componentCounts.entries
            .sortedByDescending { it.value }
            .take(5)
            .joinToString(", ") { "${it.key}(${it.value})" }

        return DetectionResult(
            patterns = patterns.toList(),
            componentCounts = componentCounts,
            summary = topComponents.ifEmpty { "No components" },
            totalNodes = tree.nodeCount
        )
    }
}

File: server/patternDetector.js
/**
 * Detect component patterns in a parsed screen tree.
 * Tags screens based on the UI components present.
 */

function detectPatterns(tree) {
  if (!tree || !tree.root) return { patterns: [], summary: 'Empty screen' };

  const patterns = [];
  const componentCounts = {};

  function walk(node) {
    if (!node) return;

    const cls = (node.class || '').toLowerCase();

    // Count component types
    const shortName = getShortName(node.class);
    componentCounts[shortName] = (componentCounts[shortName] || 0) + 1;

    // Pattern detection
    if (cls.includes('recyclerview') || cls.includes('listview')) {
      if (!patterns.includes('List')) patterns.push('List');
    }
    if (cls.includes('edittext') || cls.includes('textinputedittext') || cls.includes('textinput')) {
      if (!patterns.includes('Input Form')) patterns.push('Input Form');
    }
    if (cls.includes('bottomnavigationview') || cls.includes('bottomnavigation')) {
      if (!patterns.includes('Bottom Nav')) patterns.push('Bottom Nav');
    }
    if (cls.includes('navigationview') || cls.includes('drawerlayout')) {
      if (!patterns.includes('Navigation Drawer')) patterns.push('Navigation Drawer');
    }
    if (cls.includes('webview')) {
      if (!patterns.includes('WebView')) patterns.push('WebView');
    }
    if (cls.includes('viewpager') || cls.includes('viewpager2')) {
      if (!patterns.includes('ViewPager')) patterns.push('ViewPager');
    }
    if (cls.includes('tablayout') || cls.includes('tabhost')) {
      if (!patterns.includes('Tab Layout')) patterns.push('Tab Layout');
    }
    if (cls.includes('toolbar') || cls.includes('appbar') || cls.includes('actionbar')) {
      if (!patterns.includes('Toolbar/AppBar')) patterns.push('Toolbar/AppBar');
    }
    if (cls.includes('floatingactionbutton') || cls.includes('fab')) {
      if (!patterns.includes('FAB')) patterns.push('FAB');
    }
    if (cls.includes('imageview') || cls.includes('image')) {
      if (!patterns.includes('Images')) patterns.push('Images');
    }
    if (cls.includes('checkbox')) {
      if (!patterns.includes('Checkbox')) patterns.push('Checkbox');
    }
    if (cls.includes('radiobutton')) {
      if (!patterns.includes('Radio Button')) patterns.push('Radio Button');
    }
    if (cls.includes('switch') || cls.includes('togglebutton')) {
      if (!patterns.includes('Switch/Toggle')) patterns.push('Switch/Toggle');
    }
    if (cls.includes('spinner') || cls.includes('dropdown')) {
      if (!patterns.includes('Dropdown/Spinner')) patterns.push('Dropdown/Spinner');
    }
    if (cls.includes('searchview') || cls.includes('searchbar')) {
      if (!patterns.includes('Search Bar')) patterns.push('Search Bar');
    }
    if (cls.includes('datepicker') || cls.includes('timepicker') || cls.includes('calendar')) {
      if (!patterns.includes('Date/Time Picker')) patterns.push('Date/Time Picker');
    }
    if (cls.includes('video') || cls.includes('exoplayer') || cls.includes('mediaplayer')) {
      if (!patterns.includes('Media Player')) patterns.push('Media Player');
    }
    if (cls.includes('recyclerview') && componentCounts['RecyclerView'] > 3) {
      if (!patterns.includes('Multiple Lists')) patterns.push('Multiple Lists');
    }
    if (cls.includes('dialog') || cls.includes('alertdialog') || cls.includes('modal')) {
      if (!patterns.includes('Dialog')) patterns.push('Dialog');
    }
    if (cls.includes('progressbar') || cls.includes('loading')) {
      if (!patterns.includes('Progress/Loading')) patterns.push('Progress/Loading');
    }

    // Recurse
    if (node.children) {
      for (const child of node.children) walk(child);
    }
  }

  walk(tree.root);

  // Build summary
  const topComponents = Object.entries(componentCounts)
    .sort((a, b) => b[1] - a[1])
    .slice(0, 5)
    .map(([name, count]) => `${name}(${count})`)
    .join(', ');

  return {
    patterns,
    componentCounts,
    summary: topComponents || 'No components',
    totalNodes: tree.nodeCount
  };
}

function getShortName(className) {
  if (!className) return 'View';
  const parts = className.split('.');
  return parts[parts.length - 1] || className;
}

module.exports = { detectPatterns };
